core/lib/libtomcrypt/dsa.c

```diff
@@ -58,43 +58,35 @@ TEE_Result crypto_acipher_alloc_dsa_public_key(struct dsa_public_key *s,
 	return TEE_ERROR_OUT_OF_MEMORY;
 }
 
-TEE_Result crypto_acipher_gen_dsa_key(struct dsa_keypair *key, size_t key_size)
+TEE_Result crypto_acipher_gen_dsa_key(struct dsa_keypair *key)
 {
-	TEE_Result res;
 	dsa_key ltc_tmp_key;
-	size_t group_size, modulus_size = key_size/8;
 	int ltc_res;
 
-	if (modulus_size <= 128)
-		group_size = 20;
-	else if (modulus_size <= 256)
-		group_size = 30;
-	else if (modulus_size <= 384)
-		group_size = 35;
-	else
-		group_size = 40;
+	ltc_res = mp_init_multi(&ltc_tmp_key.g, &ltc_tmp_key.p, &ltc_tmp_key.q,
+				&ltc_tmp_key.x, &ltc_tmp_key.y, NULL);
+	if (ltc_res)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	/* Copy the key parameters */
+	mp_copy(key->g, ltc_tmp_key.g);
+	mp_copy(key->p, ltc_tmp_key.p);
+	mp_copy(key->q, ltc_tmp_key.q);
```

Comment:
Shouldn't this function check that the effective key size set in `ltc_tmp_key.p` from  matches `key_size` from caller? Maybe it is not needed as `p` defines it.

Reply:
Good question. The GP spec says "TEE_ERROR_BAD_PARAMETERS: If an incorrect or inconsistent attribute is detected. The checks that are performed depend on the implementation." and I think we could return that.
I will update and propose patches for ECC and DH who seem to suffer from the same issue.